panel/src/helpers/upload.js

```diff
@@ -1,3 +1,45 @@
+import { random } from "./string.js";
+
+/**
+ * Uploads a file in chunks
+ * @param {File} file - file to upload
+ * @param {Object} params - upload options (see `upload` method for details)
+ * @param {number} size - chunk size in bytes
+ */
+export async function chunked(file, params, size = 5242880) {
+	const parts = Math.ceil(file.size / size);
+	const id = random(4).toLowerCase();
+	let response;
+
+	for (let i = 0; i < parts; i++) {
+		const start = i * size;
+		const end = Math.min(start + size, file.size);
+		const chunk = file.slice(start, end, file.type);
+
+		// when more than one part, add flag to
+		// recognize chunked upload and its last chunk
+		if (parts > 1) {
+			params.headers["X-Upload-Chunk"] = end === file.size ? "last" : "chunk";
+		}
+
+		response = await upload(chunk, {
+			...params,
+			headers: {
+				...params.headers,
+				"X-Unique-Upload-Id": id,
+				"Content-Range": `bytes ${start}-${end - 1}/${file.size}`
```

Comment:
I think, `- 1` actually isn't needed here as `end` is already zero-indexed

Reply:
Yes, but I wouldn't use the `Content-Range` header. In requests it is [only allowed for partial `PUT` requests, with other HTTP verbs it "MUST" be ignored by the server](https://datatracker.ietf.org/doc/html/rfc9110#section-14.4-4). So we could run into situations where a web application firewall filters it out. I think that's the reason why the tus protocol uses two custom headers (see my separate review comment on the backend code).